web/src/features/AppVersionHistory/api/getVersions.jsx

```diff
@@ -0,0 +1,190 @@
+      (version.status === "failed" || version.status === "deployed");
+    const canUpdateKots =
+      version.needsKotsUpgrade && !metadata?.isAirgap && !metadata?.isKurl;
+
+    let statusLabel = "";
+
+    if (needsConfiguration) {
+      statusLabel = "Configure";
+    } else if (downstream?.currentVersion?.sequence == undefined) {
+      if (canUpdateKots) {
+        statusLabel = "Upgrade";
+      } else {
+        statusLabel = "Deploy";
+      }
+    } else if (isRedeploy) {
+      statusLabel = "Redeploy";
+    } else if (isRollback) {
+      statusLabel = "Rollback";
+    } else if (isDeploying) {
+      statusLabel = "Deploying";
+    } else if (isCurrentVersion) {
+      statusLabel = "Deployed";
+    } else {
+      if (canUpdateKots) {
+        statusLabel = "Upgrade";
+      } else {
+        statusLabel = "Deploy";
+      }
+    }
+    return {
+      version,
+      statusLabel,
+    };
+  });
+
+  return {
+    ...versions,
+    versionHistory,
+  };
+}
+
+// TODO: refactor this function so that the airgapped / nonairgapped are separate
+function getVersionsSelectorForAirgapped({ versions, currentApp, metadata }) {
+  return getVersionsSelectorForKotsManaged({ versions, currentApp, metadata });
+}
+
+function getVersionsSelectorForHelmManaged({ versions }) {
+  const deployedSequence = versions?.versionHistory?.find(
+    (v) => v.status === "deployed"
+  )?.sequence;
+
+  const versionHistory = versions?.versionHistory.map((version) => {
+    let statusLabel = "Redeploy";
+
+    if (version.sequence > deployedSequence) {
+      statusLabel = "Deploy";
+    }
+
+    if (version.sequence < deployedSequence) {
+      statusLabel = "Rollback";
+    }
+    return {
+      ...version,
+      statusLabel,
+    };
+  });
+
+  return {
+    ...versions,
+    versionHistory,
+  };
+}
+
+function chooseVersionsSelector({
+  isAirgap,
+  isKurl,
+  isHelmManaged,
+  _getVersionsSelectorForKotsManaged = getVersionsSelectorForKotsManaged,
+  _getVersionsSelectorForAirgapped = getVersionsSelectorForAirgapped,
+  _getVersionsSelectorForHelmManaged = getVersionsSelectorForHelmManaged,
+}) {
+  // if airgapped
+  if (isAirgap && isKurl) {
+    return _getVersionsSelectorForAirgapped;
+  }
+
+  // if helm managed
+  if (isHelmManaged) {
+    return _getVersionsSelectorForHelmManaged;
+  }
+
+  // if kots managed
+  return _getVersionsSelectorForKotsManaged;
+}
+
+function useVersions({
+  _getVersions = getVersions,
+  _useParams = useParams,
+  _useCurrentApp = useCurrentApp,
+  _useMetadata = useMetadata,
+  _useIsHelmManaged = useIsHelmManaged,
+} = {}) {
+  let { slug } = _useParams();
+  let { currentApp } = _useCurrentApp();
+  let { data: metadata } = _useMetadata();
+  let { data: isHelmManaged } = _useIsHelmManaged();
+
+  const versionSelector = chooseVersionsSelector({
+    // labels differ by installation manager and if airgapped
+    isAirgap: metadata?.isAirgap,
+    isHelmManaged,
+    isKurl: metadata?.isKurl,
+  });
```

Comment:
this doesn't select a version- it's a selector that's run on the `/versions` response.

Basically it determines what kind of install manage this is using and chooses the selector based on that. the Selector adds a label to the response data like "Redeploy" or "Rollback" based on the installation manager.